Update gltf Pipeline and support KHR_techniques_webgl and KHR_blend

ggetz opened the pull request:
Part of (fixes?) https://github.com/AnalyticalGraphicsInc/gltf-pipeline/issues/330
- Update `ThirdParty/gltfPipeline` files to use the [`2.0` branch](https://github.com/AnalyticalGraphicsInc/gltf-pipeline/tree/2.0)
- Support the [`KHR_techniques_webgl` extension](https://github.com/KhronosGroup/glTF/pull/1296)
- Support the [`KHR_blend` extension](https://github.com/KhronosGroup/glTF/pull/1302)
- Fixes https://github.com/AnalyticalGraphicsInc/cesium/issues/5776, removes added pipeline extras.

Ran through the Sandcastle examples to check the new updates, and the one issue I see is with classification ([in this sandcastle demo](https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/index.html?src=3D%20Tiles%20Photogrammetry%20Classification.html)). The classification model is still based on glTF 1.0, and the indices seem to be wrong. Any ideas?

Incorporates #6803 to fix bounding spheres, merge that first.

TODO:
 - [x] Include fix from https://github.com/AnalyticalGraphicsInc/cesium/pull/6857#discussion_r206157642
 - [x] Fix Specs
 - [x] Additional Specs
 - [x] Update `CHANGES.md`

cesium-concierge commented:
Thanks for the pull request @ggetz!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/gltf-update-pipeline/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Changes to third party files were made.
    * Looks like a file in one of our ThirdParty folders (`ThirdParty/, Source/ThirdParty/`) has been added or modified. Please verify that it has a section in [LICENSE.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/LICENSE.md) and that its license information is up to date with this new version.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -3354,28 +3366,28 @@ define([
+                toClipCoordinates[programId] = ModelUtility.toClipCoordinatesGLSL(model.gltf, shader);
             }
         }
 
```
Right below - `model.gltf.glExtensionsUsed` - this is now part of the techniques_webgl extension.

lilleyse commented on Source/Scene/ModelUtility.js:
```diff
@@ -33,6 +41,245 @@ define([
+     * @param {Object} gltf A javascript object containing a glTF asset.
+     * @returns {Object} The glTF asset with the added pipeline extras.
+     */
+    ModelUtility.removePipelineExtras = function(gltf) {
```
It would also be good to get around the `.extras._pipeline` usage in `splitIncompatibleMaterials`, `processPbrMetallicRoughness` and `processModelMaterialsCommon`.

lilleyse commented on Source/Scene/ModelUtility.js:
```diff
@@ -52,37 +299,37 @@ define([
-        return undefined;
+
+        return function(parameterName, attributeOrUniformName) {
+            var attributeOrUniform = technique.parameters[parameterName];
```
Is this block needed? The `KHR_techniques_webgl` extension should be defined.

lilleyse commented:
@ggetz when this is ready can you publish a new version of the power plant tileset to ion?

I checked that [upgrade](https://github.com/AnalyticalGraphicsInc/3d-tiles-tools/tree/2.0-tools/tools#upgrade) converts it correctly and it loads in this branch. That's really good because it uses pretty complicated glTF techniques and shaders.

lilleyse commented:
Related to https://github.com/AnalyticalGraphicsInc/cesium/pull/6805#discussion_r204415943, since gltf-pipeline no longer adds pipeline extras to everything there is now a bug when a model has an unused material. A few areas of the code try to access `material.extras._pipeline.primitive` when `extras._pipeline` doesn't exist.

[CUBE.gltf.txt](https://github.com/AnalyticalGraphicsInc/cesium/files/2272492/CUBE.gltf.txt)

This model reveals another bug related to `primitive.material` being undefined. I'll open a gltf-pipeline fix for that.

ggetz commented:
@lilleyse Other than specs which I'm working on now, this should be updated. I handle `ALPHA_MASK` and have the latest build of gltf-pipeline here.

I removed pipeline extras where applicable, but they are needed for `updateVersion`. Also fixed the bug related to `material.extras._pipeline.primitive`

Tested the sample models as well as Polly.

lilleyse commented:
Thanks @ggetz, I hope to look at this later today.

lilleyse commented:
The updates look good and I haven't seen any issues besides the Cesium Air ones.

I think this will cut it too close to the release and we should merge right after the release instead.

ggetz commented on Source/Scene/Model.js:
```diff
@@ -2547,209 +2399,130 @@ define([
-        rendererRenderStates[id] = RenderState.fromCache({
-            frontFace : defined(statesFunctions.frontFace) ? statesFunctions.frontFace[0] : WebGLConstants.CCW,
+        var enableCulling = !material.doubleSided;
+        var blendingEnabled = (material.alphaMode !== 'OPAQUE');
```
It still needs the depth states, I will make a separate path though for `MASK`.

lilleyse replied:
`MASK` can use the same depth states as `OPAQUE`.

lilleyse replied:
Ok I thought the PBR shader was calling `discard` if the alpha value was below the cutoff, but it looks like it isn't. Tweaking the shader to use `discard` would be preferable to enabling blending.

ggetz commented on Source/Scene/Model.js:
```diff
@@ -4242,20 +4029,20 @@ define([
-                    if (defined(this.gltf.asset) && defined(this.gltf.asset.extras) &&
-                        this.gltf.asset.extras.gltf_pipeline_upgrade_10to20) {
-                            this._forwardAxis = Axis.X;
+                    if (this.gltf.extras._pipeline.gltf_pipeline_upgrade_10to20) {
```
This was from merging in the previous forward axis fixes. I would rather do a check of the original version number before the update than update gltf-pipeline itself.

lilleyse replied:
Ah, that sounds like a better approach.

donmccurdy commented:
Just checking that I've read correctly here, is this PR intended (among other things) to let glTF-Pipeline convert from glTF 1.0 shaders to glTF 2.0 with `KHR_techniques_webgl` and `KHR_blend`? That's awesome if so, and would be very helpful to me as well. 🙂

ggetz commented:
@donmccurdy Correct, we also [move render states to KHR_BLEND](https://github.com/AnalyticalGraphicsInc/gltf-pipeline/blob/614c262a0fe25c54758b65361cfca79a0c371eef/lib/moveTechniqueRenderStates.js).

donmccurdy commented:
Is this far enough along that it could be used locally for testing output of an exporter writing `KHR_techniques_webgl`, or are there some known issues left to iron out?

ggetz commented:
@donmccurdy This should be usable for the most part. There's one bug that's causing `Cesium_Air` not to render which we are investigating, see https://github.com/AnalyticalGraphicsInc/gltf-pipeline/pull/421

ggetz commented:
@lilleyse, this should be ready for another look. I've incorporated the changes made via gltf-pipeline, fixed a bug I noticed with model caching and the forward axis, fixed the current specs, and added additional specs for loading 2.0 models and 2.0 models using the techniques extension.

lilleyse commented:
Thanks @ggetz, I will do some final testing later today.

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -4249,28 +4050,34 @@ define([
+                    if (!this._loadRendererResourcesFromCache) {
+                        var gltf = this.gltf;
+                        // Add the original version so it remains cached
+                        gltf.extras.sourceVersion = gltf.asset.version;
```
If it still works, maybe it is cleaner to store the source version as a property on model rather than extras?

lilleyse replied:
Ok, that makes sense. The method here sounds a lot easier for caching than my suggestion.

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -4249,28 +4050,34 @@ define([
+                    if (!this._loadRendererResourcesFromCache) {
+                        var gltf = this.gltf;
+                        // Add the original version so it remains cached
+                        gltf.extras.sourceVersion = gltf.asset.version;
```
I ran into an old tileset that doesn't actually set `gltf.asset` which is technically valid for gltf 1.0. The code here and in `ModelUtility` should account for that and assume `1.0` if asset is undefined.

lilleyse commented on Source/Scene/processPbrMaterials.js:
```diff
@@ -0,0 +1,786 @@
+        // so we should apply PBR unless no materials are found.
+        if (gltf.materials.length === 0) {
+            return gltf;
+        }
```
Hit a weird edge case where a gltf didn't have any meshes so it didn't even create any default materials and failed at this line. I guess a quick check that `gltf.materials` is defined should fix this. CC @OmarShehata.

lilleyse commented:
I'm seeing three test failures:

* `Scene/ModelInstanceCollection renders rigged figure`
* `Scene/ModelInstanceCollection renders with animations when instancing is disabled`
* `Scene/Model renders with a silhouette`

The last two pass when run individually.

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -251,6 +236,10 @@ define([
+     * </li><li>
+     * {@link https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_techniques_webgl/README.md|KHR_techniques_webgl}
+     * </li><li>
+     * {@link https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_blend/README.md|KHR_blend}
```
Add `KHR_materials_unlit` and `KHR_materials_pbrSpecularGlossiness` to this list. Forgot to add those in the other PRs.

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -1125,6 +1113,10 @@ define([
+     * </li><li>
+     * {@link https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_techniques_webgl/README.md|KHR_techniques_webgl}
+     * </li><li>
+     * {@link https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_blend/README.md|KHR_blend}
```
Also add those two extensions here.

lilleyse commented:
I think that's it from me. Besides those edge cases I haven't seen any other issues. 👍

ggetz commented:
Thanks @lilleyse! Tests are all cleaned up, and your feedback has been addressed.

lilleyse commented:
Thanks @ggetz. Awesome to get this in!

emackey commented:
Congrats @ggetz, @OmarShehata, and everyone involved.  It's great to have the PBR shader generation moved into core Cesium like this.
